graphbolt/include/csr_sampling_graph.h

```diff
@@ -0,0 +1,131 @@
+ * @brief Header file of csr sampling graph.
+ */
+
+#include <torch/custom_class.h>
+#include <torch/torch.h>
+
+#include <string>
+#include <vector>
+
+namespace graphbolt {
+namespace sampling {
+
+using StringList = std::vector<std::string>;
+
+/**
+ * @brief Heterogeneous information in the Graph.
+ */
+struct HeteroInfo {
+  HeteroInfo(
+      const StringList& ntypes, const StringList& etypes,
+      torch::Tensor& node_type_offset_, torch::Tensor& type_per_edge_)
+      : node_types(ntypes),
+        edge_types(etypes),
+        node_type_offset(node_type_offset_),
+        type_per_edge(type_per_edge_) {}
+  /** @brief List of node types in the graph. */
+  StringList node_types;
+  /** @brief List of edge types in the graph. */
+  StringList edge_types;
+  /** @brief Offset array of node type. */
+  torch::Tensor node_type_offset;
+  /** @brief Type of each edge. */
+  torch::Tensor type_per_edge;
+};
+
+/**
+ * @brief A sampling oriented csr format graph.
+ */
+class CSRSamplingGraph : public torch::CustomClassHolder {
+ public:
+  /**
+   * @brief Constructor for CSR with data.
+   * @param num_rows The number of rows in the dense shape.
+   * @param num_cols The number of columns in the dense shape.
+   * @param indptr The CSR format index pointer array.
+   * @param indices The CSR format index array.
+   * @param hetero_info Heterogeneous graph information, if present.
+   */
+  CSRSamplingGraph(
+      int64_t num_rows, int64_t num_cols, torch::Tensor& indptr,
+      torch::Tensor& indices, const std::shared_ptr<HeteroInfo>& hetero_info);
+
+  /**
+   * @brief Create a CSR graph from tensors of CSR format.
+   * @param indptr Index pointer array of the CSR.
+   * @param indices Indices array of the CSR.
+   * @param shape Shape of the sparse matrix.
+   *
+   * @return CSRSamplingGraph
+   */
+  static c10::intrusive_ptr<CSRSamplingGraph> FromCSR(
+      const std::vector<int64_t>& shape, torch::Tensor indptr,
+      torch::Tensor indices);
+
+  /**
+   * @brief Set the heterogeneous information for the graph.
+   *
+   * @param ntypes A list of node types.
+   * @param etypes A list of edge types.
+   * @param node_type_offset_ A tensor representing the offset of node types.
+   * @param type_per_edge_ A tensor representing the type of each edge.
+   */
+  void SetHeteroInfo(
+      const StringList& ntypes, const StringList& etypes,
+      torch::Tensor node_type_offset_, torch::Tensor type_per_edge_);
+
+  /** @brief Get the number of rows. */
+  int64_t NumRows() const { return num_rows_; }
+
+  /** @brief Get the number of columns. */
+  int64_t NumCols() const { return num_cols_; }
+
+  /** @brief Get the number of nodes. */
+  int64_t NumNodes() const { return indptr_.size(0) - 1; }
+
+  /** @brief Get the number of edges. */
+  int64_t NumEdges() const { return indices_.size(0); }
+
+  /** @brief Get the index pointer tensor. */
+  const torch::Tensor IndPtr() const { return indptr_; }
```

Comment:
Maybe better to use `CSCIndptr` to explicitly say the storage is in CSC format.

Reply:
Make sense, change for fix.

Reply:
> class name is already starts with CSC, so users should know it is a CSC graph

Make sense. 

> I may perfer IndPtr, one is to keep consistent with other member variables naming convention.

To me, `indptr` is more like a special word for CSR format. And if you use `IndPtr`, you may also need to use `ind_ptr_`. That is a bit weird... Anyway, I am not strongly insist on this.